File: BackEnd/Application/UseCases/UserGroups/RemoveUser/RemoveUserFromUserGroupCommandHandler.cs
using Domain.Aggregates;
using Infrastructure.Repository;
using Infrastructure.UnitOfWork;
using MediatR;
using Microsoft.EntityFrameworkCore;
using SharedKernel.Exceptions;

namespace Application.UseCases.UserGroups.RemoveUser;

public class RemoveUserFromUserGroupCommandHandler(
    IRepository<UserGroup, Guid> userGroupRepository,
    IUnitOfWork unitOfWork)
    : IRequestHandler<RemoveUserFromUserGroupCommand, bool>
{
    public async Task<bool> Handle(RemoveUserFromUserGroupCommand request, CancellationToken cancellationToken)
    {
        var userGroup = await userGroupRepository
            .Where(x => x.Id.Equals(request.Id))
            .Include(x => x.Users)
            .FirstOrDefaultAsync(cancellationToken);
        
        if (userGroup == null)
        {
            throw new NotFoundException("UserGroup not found");
        }

        var userToRemove = userGroup.Users.FirstOrDefault(x => x.Id == request.UserId);
        if (userToRemove == null)
        {
            throw new NotFoundException("User not found");
        }

        userGroup.Users.Remove(userToRemove);
        await userGroupRepository.UpdateAsync(userGroup);
        await unitOfWork.SaveChangesAsync(cancellationToken);
        return true;
    }
}

File: flutter/lib/application/use_cases/user_groups/remove_user_from_user_group/remove_user_from_user_group_command_handler.dart
import 'package:enterprise_management/application/use_cases/user_groups/remove_user_from_user_group/remove_user_from_user_group_command.dart';
import 'package:enterprise_management/infrastructure/repositories/user_group_repository.dart';
import 'package:enterprise_management/shared_kernel/cqrs/command_handler.dart';
import 'package:enterprise_management/shared_kernel/result/result.dart';

class RemoveUserFromUserGroupCommandHandler extends ICommandHandler<RemoveUserFromUserGroupCommand, bool> {
  const RemoveUserFromUserGroupCommandHandler({required this._userGroupRepository});

  final IUserGroupRepository _userGroupRepository;

  @override
  Future<Result<bool>> handle(RemoveUserFromUserGroupCommand command) async {
    final res = await _userGroupRepository.removeUsersFromUserGroup(id: command.id, userId: command.userId);
    return Result.success(res);
  }
}